### ISSUE TITLE
Assertions don't work in fixtures?

### ISSUE THREAD

**Issue by kevincon** [Snaipe/Criterion#211]
It seems that assertions don't work in fixtures, e.g. with this setup:

```
static void my_suite_teardown(void)
{
	cr_assert(false);
}

TestSuite(my_suite, .fini = my_suite_teardown);

Test(my_suite, my_test)
{
	cr_assert(true);
}
```

I get the following output when running the test binary:

```
[====] Running 1 test from my_suite:
[RUN ] my_suite::my_test
[PASS] my_suite::my_test: (0.00s)
[----] /path/to/my_test.c:1337: Assertion failed: The expression 0 is false.
[----] Warning! The test `my_suite::my_test` crashed during its setup or teardown.
[====] Synthesis: Tested: 1 | Passing: 1 | Failing: 0 | Crashing: 0
```

So since the failed assertion does not occur in the test case function, the test is still recorded as passing but we also see the `Assertion failed` message and a warning that the test crashed during its teardown. Although surprisingly, the synthesis shows `Crashing: 0`. I also got the same result when trying a fixture at the test case level.

Would it be possible to support assertions in fixtures to where a failing assertion in a fixture fails the test case being run? I've found it useful in other test frameworks like Google Test to be able to validate that all of the test cases in a suite have met some expectation, e.g. that some resource has been completely cleaned up, without duplicating that expectation in each test case.

Or alternatively (if it's more feasible), could crashing using an assertion like this be recorded in the synthesis as "Crashing" so that the test binary returns a non-zero error code? That would allow my helper script to still recognize that a test binary failed.

**Comment by kevincon** [comment]
I also just tried asserting in a `.init` fixture and interestingly that does mark the test as failed as well as crashing. So maybe it's just the `.fini` fixtures that aren't being marked as failed or crashing?

**Comment by Snaipe** [comment]
Well, assertions here are working half-heartedly. They do abort the test on failure, but since we're not actually crashing, they just don't do anything in this instance. The warning message here is misleading.

Currently, criterion assumes that everything related to strictly testing is done in the body of the test. Code that run in init/fini is therefore strictly for initialization and finalization, and are expected to work as part of the contract of the test.

Hence, if a test dies for any reason in .init, then the test fails, because you can't fulfil the conditions to run the tests. However, if .fini dies, even though the test was reported as passing, then this is a problem. It's not completely horrible since, well, the test passed, and the process was going to die anyway, so ressource collection is done by the kernel, but we still ought to warn it.

Perhaps a teardown fixture that fail should mark the test as failed, but it would make it hard to distinguish between tests that failed *and* that died during teardown, and tests that passed and died.

In any case, yes, warnings should return a nonzero exit status (#169), and the warning should definitely be clarified.

**Comment by Snaipe** [comment]
This should be fixed by 24005f4.
